### ISSUE TITLE
Feedback: Saga Timeouts - NServiceBus

### ISSUE THREAD

**Issue by Niklas-Peter** [Particular/docs.particular.net#2523]
Page: https://docs.particular.net/nservicebus/sagas/timeouts

> If a saga is recreated based on a similar message key then this is not the same saga instance and this old timeout message will not be processed by this new saga instance that shares the same key. The timeout message will be ignored.

> NServiceBus will make sure that all properties used for correlation are unique across all instances of the given saga type. How this is enforced is up to each persister but will most likely translate to a unique key constraint in the database.

(https://docs.particular.net/nservicebus/sagas/message-correlation)

> MarkAsComplete(): This may result in the sagas state being deleted by the persister.

(https://github.com/Particular/NServiceBus/blob/18e4f48f34553206fc9ae9697ac12d93c19d6659/src/NServiceBus.Core/Sagas/Saga.cs#L126)

Is it safe to complete a saga instance with correlation key X and immediately start another saga instance with the same key X? I use a saga for aggregating messages and I complete the  saga as soon as a specific number N of messages was aggregated. However, after the first N messages arrived, further messages starting a saga instance with the same correlation ID might be received.

If `MarkAsComplete()` only _may_ delete the state and at the same time the correlation ID has a unique constraint in the persistence, you would get a duplicate exception.

So far I could not detect this problem in the implementation. I just concluded that from your documentation.

**Comment by indualagarsamy** [comment]
@Niklas-Peter - Thanks for your question.  Calling MarkAsComplete() will delete the saga entry from the persistence. When a saga is completed and you have a new message with the same correlation id, that should start a new saga instance. 
I think the code comment needs to get updated. 

@Particular/nservicebus-maintainers - please review and comment: https://github.com/Particular/NServiceBus/blob/18e4f48f34553206fc9ae9697ac12d93c19d6659/src/NServiceBus.Core/Sagas/Saga.cs#L126

**Comment by timbussmann** [comment]
@indualagarsamy is correct that there will be a new saga instance created. As for the comment on the method I'm on the fence. It's up to the persistence whether they delete the saga instance data or not, but even if the saga isn't deleted it would not interfere with the mentioned behavior, as each saga instance has an unique id (not the correlation id) which maps to the correlation id. If a saga is completed, the completed saga will not be found when a message with the same correlation id arrives but a new saga instance with a new saga id will be created.

Suggestions how that code comment could be changed? I think we should remove any statement about the persisters behavior?

**Comment by indualagarsamy** [comment]
@timbussmann - For: https://github.com/Particular/NServiceBus/blob/18e4f48f34553206fc9ae9697ac12d93c19d6659/src/NServiceBus.Core/Sagas/Saga.cs#L126

Instead of: 

>This may result in the sagas state being deleted by the persister.

How about: 

This will result in the saga being marked as closed by the persistence. Any pending timeouts for the current saga instance will be removed. In the event of a race condition that the timeout message was just dispatched, when the saga is being marked as complete, that timeout message will be ignored.

**Comment by andreasohlund** [comment]
None of the transports that have native timeout support are able to "remove timeouts by saga id" so I think we have to see that more and more as an under the hood optimization?

**Comment by timbussmann** [comment]
> This will result in the saga being marked as closed by the persistence.

I don't like the term closed. From an API perspective we call it "completed" and neither of them explains what this means in technical terms.

I'm with andreas here, this sounds like a too detailed description of a potential implementation. Also, what about responses? It's not just replies I think.

**Comment by SimonCropp** [comment]
@timbussmann @andreasohlund can u submit a PR with your proposal

**Comment by weralabaj** [comment]
@timbussmann @andreasohlund ping ☝️

**Comment by timbussmann** [comment]
I have no capacity to address this atm.

**Comment by andreasohlund** [comment]
>Is it safe to complete a saga instance with correlation key X and immediately start another saga instance with the same key X? I use a saga for aggregating messages and I complete the saga as soon as a specific number N of messages was aggregated. However, after the first N messages arrived, further messages starting a saga instance with the same correlation ID might be received.

Since we can't guarantee when and even if timeouts will be discarded its not supported to open a new saga with the same ID as one that have been closed. The workaround is to keep the saga open forever.

Do we all agree to that statement?

**Comment by andreasohlund** [comment]
>Since we can't guarantee when and even if timeouts will be discarded its not supported to open a new saga with the same ID as one that have been closed

I guess we can add "if you care about timeouts"?

Writing that makes me realize that it would all work as long as you use the timeout state to tag the timeouts to the correct saga (timestamp etc).

Think it might still be issues since we don't require the saga instance to be removed from storage immediately so the behavior would be different depending if the storage has been able to clean up the instance or not.

**Comment by weralabaj** [comment]
@andreasohlund Do you need some help with this? not sure how to push it forward.

**Comment by weralabaj** [comment]
Hi @Niklas-Peter !

Sorry for the delay, we dropped a ball here. I discussed this problem with @andreasohlund and seems like we can find technical workarounds for the problem, but it gets complicated really quickly.

This
> However, after the first N messages arrived, further messages starting a saga instance with the same correlation ID might be received.

made me wonder what's your scenario from the business perspective? Is there any chance to avoid this situation (having two sagas with the same correlation ID)? We're inclined to suggest trying to modify the business model in a way that avoids the problem in the first place, as technical solutions will be tricky.

I'd be happy to jump on a call or open a support ticket to discuss this and learn more about this scenario if you want.
